{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "pd.options.display.max_colwidth=100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def jaccard_similarity(str1, str2):\n",
    "    string_set1 = set(str1.split(' '))\n",
    "    string_set2 = set(str2.split(' '))\n",
    "    return len(string_set1.intersection(string_set2))/len(string_set2.union(string_set1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dump pkl into a pandas dataframe. \n",
    "data = pd.read_pickle('data/string_study.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"word_count\"] = data._info.apply(lambda x:len(x.split(' ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "class_by_len = set(data[\"word_count\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# class_by_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "45"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bin_len = len(class_by_len)\n",
    "bin_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x121864be0>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAD8CAYAAACcjGjIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAE7tJREFUeJzt3X+s3fV93/HnazgkDNIYQndlARpEtTq5ZSNgAVGi6ibR+JWpUCmKQKh4KaurBaREQ1pMp40uJBKZRrqBUlq38QISDWH5MSPijHmUqyqTIEBCMD9KcYkjbAFeYn7USdXM2Xt/nI+TU3/uxZdzj+/9Zn4+pK/u93y+P87r6wN+3e/3fM9xqgpJksb9vZUOIEkaHstBktSxHCRJHctBktSxHCRJHctBktSxHCRJHctBktSxHCRJnVUrHWBSJ598cp1++ukTbfvDH/6Q448/frqBpmjo+cCM0zD0fGDGaRhavkcfffT7VfWLh12xqn4up3POOacm9cADD0y87XIYer4qM07D0PNVmXEahpYPeKQW8Xesl5UkSR3LQZLUsRwkSR3LQZLUsRwkSR3LQZLUsRwkSR3LQZLUsRwkSZ2f26/PWIode17ln2/62rzLdt30gWVOI0nD45mDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOocthySnJXkgyVNJnkzy0Tb+e0n2JHmsTZeMbXN9kp1Jnkly4dj4RW1sZ5JNY+NnJHmojX8xybHTPlBJ0uIt5szhAHBdVa0DzgeuSbKuLfv9qjqrTdsA2rLLgV8BLgL+IMkxSY4BPgtcDKwDrhjbz6fbvn4JeBm4ekrHJ0mawGHLoapeqKpvtfm/Bp4GTnmdTS4F7qqqv62q7wI7gXPbtLOqnquqHwN3AZcmCfA+4Ett+9uByyY9IEnS0r2h9xySnA68E3ioDV2b5PEkW5Kc2MZOAZ4f22x3G1to/O3AK1V14JBxSdIKWfRXdic5Afgy8LGqei3JbcCNQLWfNwO/dURS/izDRmAjwMzMDHNzcxPtZ+Y4uO7MA/Mum3Sf07R///5B5Hg9Zly6oecDM07D0PMtZFHlkORNjIrhzqr6CkBVvTS2/I+Be9vDPcBpY5uf2sZYYPwHwOokq9rZw/j6f0dVbQY2A6xfv75mZ2cXE79z651buXnH/Ie+68rJ9jlNc3NzTHpsy8WMSzf0fGDGaRh6voUs5m6lAJ8Dnq6qz4yNrxlb7TeAJ9r8PcDlSd6c5AxgLfBN4GFgbbsz6VhGb1rfU1UFPAB8sG2/Adi6tMOSJC3FYs4c3g38JrAjyWNt7HcZ3W10FqPLSruA3wGoqieT3A08xehOp2uq6icASa4F7gOOAbZU1ZNtfx8H7krySeDbjMpIkrRCDlsOVfUNIPMs2vY623wK+NQ849vm266qnmN0N5MkaQD8hLQkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqWM5SJI6loMkqXPYckhyWpIHkjyV5MkkH23jJyXZnuTZ9vPENp4ktyTZmeTxJGeP7WtDW//ZJBvGxs9JsqNtc0uSHImDlSQtzmLOHA4A11XVOuB84Jok64BNwP1VtRa4vz0GuBhY26aNwG0wKhPgBuA84FzghoOF0tb57bHtLlr6oUmSJnXYcqiqF6rqW23+r4GngVOAS4Hb22q3A5e1+UuBO2rkQWB1kjXAhcD2qtpXVS8D24GL2rJfqKoHq6qAO8b2JUlaAW/oPYckpwPvBB4CZqrqhbboRWCmzZ8CPD+22e429nrju+cZlyStkFWLXTHJCcCXgY9V1WvjbwtUVSWpI5Dv0AwbGV2qYmZmhrm5uYn2M3McXHfmgXmXTbrPadq/f/8gcrweMy7d0POBGadh6PkWsqhySPImRsVwZ1V9pQ2/lGRNVb3QLg3tbeN7gNPGNj+1je0BZg8Zn2vjp86zfqeqNgObAdavX1+zs7PzrXZYt965lZt3zH/ou66cbJ/TNDc3x6THtlzMuHRDzwdmnIah51vIYu5WCvA54Omq+szYonuAg3ccbQC2jo1f1e5aOh94tV1+ug+4IMmJ7Y3oC4D72rLXkpzfnuuqsX1JklbAYs4c3g38JrAjyWNt7HeBm4C7k1wNfA/4UFu2DbgE2An8CPgwQFXtS3Ij8HBb7xNVta/NfwT4PHAc8PU2SZJWyGHLoaq+ASz0uYP3z7N+AdcssK8twJZ5xh8BfvVwWSRJy8NPSEuSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKlz2HJIsiXJ3iRPjI39XpI9SR5r0yVjy65PsjPJM0kuHBu/qI3tTLJpbPyMJA+18S8mOXaaByhJeuMWc+bweeCiecZ/v6rOatM2gCTrgMuBX2nb/EGSY5IcA3wWuBhYB1zR1gX4dNvXLwEvA1cv5YAkSUt32HKoqj8H9i1yf5cCd1XV31bVd4GdwLlt2llVz1XVj4G7gEuTBHgf8KW2/e3AZW/wGCRJU7ZqCdtem+Qq4BHguqp6GTgFeHBsnd1tDOD5Q8bPA94OvFJVB+ZZv5NkI7ARYGZmhrm5uYmCzxwH1515YN5lk+5zmvbv3z+IHK/HjEs39HxgxmkYer6FTFoOtwE3AtV+3gz81rRCLaSqNgObAdavX1+zs7MT7efWO7dy8475D33XlZPtc5rm5uaY9NiWixmXbuj5wIzTMPR8C5moHKrqpYPzSf4YuLc93AOcNrbqqW2MBcZ/AKxOsqqdPYyvL0laIRPdyppkzdjD3wAO3sl0D3B5kjcnOQNYC3wTeBhY2+5MOpbRm9b3VFUBDwAfbNtvALZOkkmSND2HPXNI8gVgFjg5yW7gBmA2yVmMLivtAn4HoKqeTHI38BRwALimqn7S9nMtcB9wDLClqp5sT/Fx4K4knwS+DXxuakcnSZrIYcuhqq6YZ3jBv8Cr6lPAp+YZ3wZsm2f8OUZ3M0mSBsJPSEuSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOpaDJKljOUiSOocthyRbkuxN8sTY2ElJtid5tv08sY0nyS1JdiZ5PMnZY9tsaOs/m2TD2Pg5SXa0bW5JkmkfpCTpjVnMmcPngYsOGdsE3F9Va4H722OAi4G1bdoI3AajMgFuAM4DzgVuOFgobZ3fHtvu0OeSJC2zw5ZDVf05sO+Q4UuB29v87cBlY+N31MiDwOoka4ALge1Vta+qXga2Axe1Zb9QVQ9WVQF3jO1LkrRCJn3PYaaqXmjzLwIzbf4U4Pmx9Xa3sdcb3z3PuCRpBa1a6g6qqpLUNMIcTpKNjC5XMTMzw9zc3ET7mTkOrjvzwLzLJt3nNO3fv38QOV6PGZdu6PnAjNMw9HwLmbQcXkqypqpeaJeG9rbxPcBpY+ud2sb2ALOHjM+18VPnWX9eVbUZ2Aywfv36mp2dXWjV13XrnVu5ecf8h77rysn2OU1zc3NMemzLxYxLN/R8YMZpGHq+hUx6Weke4OAdRxuArWPjV7W7ls4HXm2Xn+4DLkhyYnsj+gLgvrbstSTnt7uUrhrblyRphRz2zCHJFxj91n9ykt2M7jq6Cbg7ydXA94APtdW3AZcAO4EfAR8GqKp9SW4EHm7rfaKqDr7J/RFGd0QdB3y9TZKkFXTYcqiqKxZY9P551i3gmgX2swXYMs/4I8CvHi6HJGn5+AlpSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdSwHSVJnyf+G9P9vTt/0tXnHd930gWVOIkkrxzMHSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdSwHSVLHcpAkdZZUDkl2JdmR5LEkj7Sxk5JsT/Js+3liG0+SW5LsTPJ4krPH9rOhrf9skg1LOyRJ0lJN48zhvVV1VlWtb483AfdX1Vrg/vYY4GJgbZs2ArfBqEyAG4DzgHOBGw4WiiRpZRyJy0qXAre3+duBy8bG76iRB4HVSdYAFwLbq2pfVb0MbAcuOgK5JEmLlKqafOPku8DLQAF/VFWbk7xSVavb8gAvV9XqJPcCN1XVN9qy+4GPA7PAW6rqk2383wJ/U1X/cZ7n28jorIOZmZlz7rrrroly7933Ki/9zRvb5sxT3jbRc01i//79nHDCCcv2fJMw49INPR+YcRqGlu+9733vo2NXeha01H8m9D1VtSfJPwC2J/mL8YVVVUkmb59DVNVmYDPA+vXra3Z2dqL93HrnVm7e8cYOfdeVkz3XJObm5pj02JaLGZdu6PnAjNMw9HwLWdJlpara037uBb7K6D2Dl9rlItrPvW31PcBpY5uf2sYWGpckrZCJyyHJ8UneenAeuAB4ArgHOHjH0QZga5u/B7iq3bV0PvBqVb0A3AdckOTE9kb0BW1MkrRClnJZaQb46uhtBVYBf1pV/z3Jw8DdSa4Gvgd8qK2/DbgE2An8CPgwQFXtS3Ij8HBb7xNVtW8JuSRJSzRxOVTVc8A/mWf8B8D75xkv4JoF9rUF2DJpFknSdPkJaUlSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSx3KQJHUsB0lSZzDlkOSiJM8k2Zlk00rnkaSj2SDKIckxwGeBi4F1wBVJ1q1sKkk6eg2iHIBzgZ1V9VxV/Ri4C7h0hTNJ0lFr1UoHaE4Bnh97vBs4b4WyzOv0TV9bcNmumz6wjEkk6cgbSjksSpKNwMb2cH+SZybc1cnA96eTCvLpae3pp6aa7wgx49INPR+YcRqGlu8fLmaloZTDHuC0scentrG/o6o2A5uX+mRJHqmq9Uvdz5Ey9HxgxmkYej4w4zQMPd9ChvKew8PA2iRnJDkWuBy4Z4UzSdJRaxBnDlV1IMm1wH3AMcCWqnpyhWNJ0lFrEOUAUFXbgG3L9HRLvjR1hA09H5hxGoaeD8w4DUPPN69U1UpnkCQNzFDec5AkDchRVQ5D+YqOJFuS7E3yxNjYSUm2J3m2/TyxjSfJLS3z40nOXoZ8pyV5IMlTSZ5M8tEBZnxLkm8m+U7L+O/b+BlJHmpZvthucCDJm9vjnW356Uc6Y3veY5J8O8m9A823K8mOJI8leaSNDeZ1bs+7OsmXkvxFkqeTvGtIGZP8cvvzOzi9luRjQ8o4kao6KiZGb3T/FfAO4FjgO8C6Fcrya8DZwBNjY/8B2NTmNwGfbvOXAF8HApwPPLQM+dYAZ7f5twJ/yehrTYaUMcAJbf5NwEPtue8GLm/jfwj8yzb/EeAP2/zlwBeX6bX+V8CfAve2x0PLtws4+ZCxwbzO7XlvB/5Fmz8WWD20jGNZjwFeZPRZgkFmXPSxrHSAZXzR3gXcN/b4euD6Fcxz+iHl8Aywps2vAZ5p838EXDHfesuYdSvwT4eaEfj7wLcYfar++8CqQ19zRnfCvavNr2rr5QjnOhW4H3gfcG/7y2Aw+dpzzVcOg3mdgbcB3z30z2JIGQ/JdQHwv4accbHT0XRZab6v6DhlhbLMZ6aqXmjzLwIzbX5Fc7fLG+9k9Jv5oDK2SzaPAXuB7YzODF+pqgPz5Phpxrb8VeDtRzjifwL+NfB/2+O3DywfQAH/I8mjGX0DAQzrdT4D+N/Af2mX5/4kyfEDyzjucuALbX6oGRflaCqHnxs1+nVixW8jS3IC8GXgY1X12viyIWSsqp9U1VmMfkM/F/hHK5lnXJJ/BuytqkdXOsthvKeqzmb0jcjXJPm18YUDeJ1XMboEe1tVvRP4IaNLND81gIwAtPePfh34r4cuG0rGN+JoKodFfUXHCnopyRqA9nNvG1+R3EnexKgY7qyqrwwx40FV9QrwAKPLNKuTHPz8zniOn2Zsy98G/OAIxno38OtJdjH6luH3Af95QPkAqKo97ede4KuMSnZIr/NuYHdVPdQef4lRWQwp40EXA9+qqpfa4yFmXLSjqRyG/hUd9wAb2vwGRtf5D45f1e5wOB94dexU9YhIEuBzwNNV9ZmBZvzFJKvb/HGM3hN5mlFJfHCBjAezfxD4s/bb3BFRVddX1alVdTqj/9b+rKquHEo+gCTHJ3nrwXlG18ufYECvc1W9CDyf5Jfb0PuBp4aUccwV/OyS0sEsQ8u4eCv9psdyTozuEvhLRtem/80K5vgC8ALwfxj9ZnQ1o+vL9wPPAv8TOKmtG0b/ENJfATuA9cuQ7z2MToEfBx5r0yUDy/iPgW+3jE8A/66NvwP4JrCT0en9m9v4W9rjnW35O5bx9Z7lZ3crDSZfy/KdNj158P+JIb3O7XnPAh5pr/V/A04cYMbjGZ3pvW1sbFAZ3+jkJ6QlSZ2j6bKSJGmRLAdJUsdykCR1LAdJUsdykCR1LAdJUsdykCR1LAdJUuf/AdhnJEzuISjeAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Analysis of messages categorized into words\n",
    "data.word_count.hist(bins = bin_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the class with highest freq. \n",
    "dict = {}\n",
    "for item in class_by_len:\n",
    "    temp_frame = np.array(data.query(\"word_count == @item\"))\n",
    "    dict[str(item)] = len(temp_frame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>atime</th>\n",
       "      <th>task</th>\n",
       "      <th>deployment</th>\n",
       "      <th>raw</th>\n",
       "      <th>dtime</th>\n",
       "      <th>msg</th>\n",
       "      <th>_info</th>\n",
       "      <th>word_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9162</th>\n",
       "      <td>1522653718630</td>\n",
       "      <td>boot-linux</td>\n",
       "      <td>wig_project_014</td>\n",
       "      <td>2018-04-02 09:21:58.630 21399 ERROR rallytester.rallytester [-] [nova boot-linux wig_project_014...</td>\n",
       "      <td>2018-04-02 07:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-e9e5-VJdA&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20215</th>\n",
       "      <td>1524045706645</td>\n",
       "      <td>boot-from-snapshot-linux</td>\n",
       "      <td>gva_project_035</td>\n",
       "      <td>2018-04-18 12:01:46.645 17144 ERROR rallytester.rallytester [-] [nova boot-from-snapshot-linux g...</td>\n",
       "      <td>2018-04-18 10:00:00</td>\n",
       "      <td>The server is currently unavailable. Please try again at a later time.&lt;br /&gt;&lt;br /&gt;    (HTTP 503)...</td>\n",
       "      <td>The server is currently unavailable. Please try again at a later time.&lt;br /&gt;&lt;br /&gt;    (HTTP 503)...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27306</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27339</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27354</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27399</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27453</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27582</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27690</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27762</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27787</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27825</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27881</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27907</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27914</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27915</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27953</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28010</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28025</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28146</th>\n",
       "      <td>1522949622361</td>\n",
       "      <td>boot-from-snapshot-linux</td>\n",
       "      <td>gva_project_032</td>\n",
       "      <td>2018-04-05 19:33:42.361 23970 ERROR rallytester.rallytester [-] [nova boot-from-snapshot-linux g...</td>\n",
       "      <td>2018-04-05 17:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-2150-kVPS&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28149</th>\n",
       "      <td>1522916259808</td>\n",
       "      <td>attach-volume</td>\n",
       "      <td>gva_project_013</td>\n",
       "      <td>2018-04-05 10:17:39.808 7761 ERROR rallytester.rallytester [-] [nova attach-volume gva_project_0...</td>\n",
       "      <td>2018-04-05 08:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-9272-P7n5&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28341</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28350</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28400</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28478</th>\n",
       "      <td>1522919580867</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>qa</td>\n",
       "      <td>2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...</td>\n",
       "      <td>2018-04-05 09:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-0d45-N5Bh&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28657</th>\n",
       "      <td>1522951319527</td>\n",
       "      <td>attach-volume</td>\n",
       "      <td>gva_project_035</td>\n",
       "      <td>2018-04-05 20:01:59.527 25597 ERROR rallytester.rallytester [-] [nova attach-volume gva_project_...</td>\n",
       "      <td>2018-04-05 18:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-f1c8-hEwi&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28664</th>\n",
       "      <td>1522957909078</td>\n",
       "      <td>attach-volume</td>\n",
       "      <td>wig_project_006</td>\n",
       "      <td>2018-04-05 21:51:49.078 26798 ERROR rallytester.rallytester [-] [nova attach-volume wig_project_...</td>\n",
       "      <td>2018-04-05 19:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-261f-Bn42&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28665</th>\n",
       "      <td>1522957909078</td>\n",
       "      <td>attach-volume</td>\n",
       "      <td>wig_project_006</td>\n",
       "      <td>2018-04-05 21:51:49.078 26798 ERROR rallytester.rallytester [-] [nova attach-volume wig_project_...</td>\n",
       "      <td>2018-04-05 19:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-261f-Bn42&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30694</th>\n",
       "      <td>1522977882448</td>\n",
       "      <td>ping-linux</td>\n",
       "      <td>gva_project_035</td>\n",
       "      <td>2018-04-06 03:24:42.448 23870 ERROR rallytester.rallytester [-] [nova ping-linux gva_project_035...</td>\n",
       "      <td>2018-04-06 01:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-798e-7yAV&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32047</th>\n",
       "      <td>1522625408426</td>\n",
       "      <td>boot-from-snapshot-linux</td>\n",
       "      <td>gva_project_026</td>\n",
       "      <td>2018-04-02 01:30:08.426 28739 ERROR rallytester.rallytester [-] [nova boot-from-snapshot-linux g...</td>\n",
       "      <td>2018-04-01 23:00:00</td>\n",
       "      <td>Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))</td>\n",
       "      <td>Resource &lt;Server: rally-cdec-Ujlg&gt; has ERROR status. Fault: {u'message': u'Unable to establish c...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32525</th>\n",
       "      <td>1523430208984</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>wig_project_009</td>\n",
       "      <td>2018-04-11 09:03:28.984 11899 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-66b8-K4E5&gt; has ERROR status. Fault: {u'message': u'Build of instance 860...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32526</th>\n",
       "      <td>1523430206792</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>gva_shared_007</td>\n",
       "      <td>2018-04-11 09:03:26.792 11507 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-efaf-125N&gt; has ERROR status. Fault: {u'message': u'Build of instance f7b...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32528</th>\n",
       "      <td>1523430227607</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>gva_shared_006</td>\n",
       "      <td>2018-04-11 09:03:47.607 11675 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-af3e-i2Yw&gt; has ERROR status. Fault: {u'message': u'Build of instance 280...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32555</th>\n",
       "      <td>1523430202922</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>gva_shared_012</td>\n",
       "      <td>2018-04-11 09:03:22.922 10892 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-f865-WR2B&gt; has ERROR status. Fault: {u'message': u'Build of instance a25...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32585</th>\n",
       "      <td>1523430207143</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>pre_stage</td>\n",
       "      <td>2018-04-11 09:03:27.143 12129 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux pre...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-76ab-4IZk&gt; has ERROR status. Fault: {u'message': u'Build of instance 5e9...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32586</th>\n",
       "      <td>1523430233673</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>gva_shared_002</td>\n",
       "      <td>2018-04-11 09:03:53.673 6820 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva_...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-1fe3-msFb&gt; has ERROR status. Fault: {u'message': u'Build of instance 75b...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32587</th>\n",
       "      <td>1523430240602</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>gva_project_019</td>\n",
       "      <td>2018-04-11 09:04:00.602 9952 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva_...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-8c5a-c7MX&gt; has ERROR status. Fault: {u'message': u'Build of instance 556...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32588</th>\n",
       "      <td>1523430246481</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>wig_project_008</td>\n",
       "      <td>2018-04-11 09:04:06.481 8669 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig_...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-1e6a-twpC&gt; has ERROR status. Fault: {u'message': u'Build of instance eb6...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32626</th>\n",
       "      <td>1523430210221</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>gva_project_041</td>\n",
       "      <td>2018-04-11 09:03:30.221 9586 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva_...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-e1fe-ahg2&gt; has ERROR status. Fault: {u'message': u'Build of instance 96b...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32627</th>\n",
       "      <td>1523430219701</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>wig_project_012</td>\n",
       "      <td>2018-04-11 09:03:39.701 7214 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig_...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-6490-afBB&gt; has ERROR status. Fault: {u'message': u'Build of instance f91...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32628</th>\n",
       "      <td>1523430229423</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>gva_project_017</td>\n",
       "      <td>2018-04-11 09:03:49.423 6221 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva_...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-6d6b-dOgn&gt; has ERROR status. Fault: {u'message': u'Build of instance 5c1...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32630</th>\n",
       "      <td>1523430241291</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>gva_project_018</td>\n",
       "      <td>2018-04-11 09:04:01.291 12322 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-0f56-eXg7&gt; has ERROR status. Fault: {u'message': u'Build of instance 43e...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32631</th>\n",
       "      <td>1523430246266</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>crit_project_001</td>\n",
       "      <td>2018-04-11 09:04:06.266 6822 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux crit...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-186e-Tfnx&gt; has ERROR status. Fault: {u'message': u'Build of instance 6a3...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32644</th>\n",
       "      <td>1523430214322</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>wig_project_002</td>\n",
       "      <td>2018-04-11 09:03:34.322 12603 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-ef1c-55VR&gt; has ERROR status. Fault: {u'message': u'Build of instance 1c0...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32645</th>\n",
       "      <td>1523430237825</td>\n",
       "      <td>boot-from-volume-linux</td>\n",
       "      <td>wig_project_006</td>\n",
       "      <td>2018-04-11 09:03:57.825 10721 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig...</td>\n",
       "      <td>2018-04-11 07:00:00</td>\n",
       "      <td>Build of instance xyz aborted: Service Unavailable (HTTP 503)</td>\n",
       "      <td>Resource &lt;Server: rally-6e4b-ocjj&gt; has ERROR status. Fault: {u'message': u'Build of instance f87...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32810</th>\n",
       "      <td>1523333412352</td>\n",
       "      <td>ping-linux</td>\n",
       "      <td>gva_project_038</td>\n",
       "      <td>2018-04-10 06:10:12.352 3761 ERROR rallytester.rallytester [-] [nova ping-linux gva_project_038]...</td>\n",
       "      <td>2018-04-10 04:00:00</td>\n",
       "      <td>Script execution failed: Error running command {\"interpreter\": \"/bin/sh\", \"script_inline\": \"ping...</td>\n",
       "      <td>Script execution failed: Error running command {\"interpreter\": \"/bin/sh\", \"script_inline\": \"ping...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33136</th>\n",
       "      <td>1522831217039</td>\n",
       "      <td>reboot-linux</td>\n",
       "      <td>gva_project_001</td>\n",
       "      <td>2018-04-04 10:40:17.039 12381 ERROR rallytester.rallytester [-] [nova reboot-linux gva_project_0...</td>\n",
       "      <td>2018-04-04 08:00:00</td>\n",
       "      <td>Connection to the hypervisor is broken on host: xyz.cern.ch</td>\n",
       "      <td>Resource &lt;Server: rally-9f83-Kq8d&gt; has ERROR status. Fault: {u'message': u'Connection to the hyp...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               atime                      task        deployment  \\\n",
       "9162   1522653718630                boot-linux   wig_project_014   \n",
       "20215  1524045706645  boot-from-snapshot-linux   gva_project_035   \n",
       "27306  1522919580867              reboot-linux                qa   \n",
       "27339  1522919580867              reboot-linux                qa   \n",
       "27354  1522919580867              reboot-linux                qa   \n",
       "27399  1522919580867              reboot-linux                qa   \n",
       "27453  1522919580867              reboot-linux                qa   \n",
       "27582  1522919580867              reboot-linux                qa   \n",
       "27690  1522919580867              reboot-linux                qa   \n",
       "27762  1522919580867              reboot-linux                qa   \n",
       "27787  1522919580867              reboot-linux                qa   \n",
       "27825  1522919580867              reboot-linux                qa   \n",
       "27881  1522919580867              reboot-linux                qa   \n",
       "27907  1522919580867              reboot-linux                qa   \n",
       "27914  1522919580867              reboot-linux                qa   \n",
       "27915  1522919580867              reboot-linux                qa   \n",
       "27953  1522919580867              reboot-linux                qa   \n",
       "28010  1522919580867              reboot-linux                qa   \n",
       "28025  1522919580867              reboot-linux                qa   \n",
       "28146  1522949622361  boot-from-snapshot-linux   gva_project_032   \n",
       "28149  1522916259808             attach-volume   gva_project_013   \n",
       "28341  1522919580867              reboot-linux                qa   \n",
       "28350  1522919580867              reboot-linux                qa   \n",
       "28400  1522919580867              reboot-linux                qa   \n",
       "28478  1522919580867              reboot-linux                qa   \n",
       "28657  1522951319527             attach-volume   gva_project_035   \n",
       "28664  1522957909078             attach-volume   wig_project_006   \n",
       "28665  1522957909078             attach-volume   wig_project_006   \n",
       "30694  1522977882448                ping-linux   gva_project_035   \n",
       "32047  1522625408426  boot-from-snapshot-linux   gva_project_026   \n",
       "32525  1523430208984    boot-from-volume-linux   wig_project_009   \n",
       "32526  1523430206792    boot-from-volume-linux    gva_shared_007   \n",
       "32528  1523430227607    boot-from-volume-linux    gva_shared_006   \n",
       "32555  1523430202922    boot-from-volume-linux    gva_shared_012   \n",
       "32585  1523430207143    boot-from-volume-linux         pre_stage   \n",
       "32586  1523430233673    boot-from-volume-linux    gva_shared_002   \n",
       "32587  1523430240602    boot-from-volume-linux   gva_project_019   \n",
       "32588  1523430246481    boot-from-volume-linux   wig_project_008   \n",
       "32626  1523430210221    boot-from-volume-linux   gva_project_041   \n",
       "32627  1523430219701    boot-from-volume-linux   wig_project_012   \n",
       "32628  1523430229423    boot-from-volume-linux   gva_project_017   \n",
       "32630  1523430241291    boot-from-volume-linux   gva_project_018   \n",
       "32631  1523430246266    boot-from-volume-linux  crit_project_001   \n",
       "32644  1523430214322    boot-from-volume-linux   wig_project_002   \n",
       "32645  1523430237825    boot-from-volume-linux   wig_project_006   \n",
       "32810  1523333412352                ping-linux   gva_project_038   \n",
       "33136  1522831217039              reboot-linux   gva_project_001   \n",
       "\n",
       "                                                                                                       raw  \\\n",
       "9162   2018-04-02 09:21:58.630 21399 ERROR rallytester.rallytester [-] [nova boot-linux wig_project_014...   \n",
       "20215  2018-04-18 12:01:46.645 17144 ERROR rallytester.rallytester [-] [nova boot-from-snapshot-linux g...   \n",
       "27306  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27339  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27354  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27399  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27453  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27582  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27690  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27762  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27787  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27825  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27881  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27907  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27914  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27915  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "27953  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "28010  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "28025  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "28146  2018-04-05 19:33:42.361 23970 ERROR rallytester.rallytester [-] [nova boot-from-snapshot-linux g...   \n",
       "28149  2018-04-05 10:17:39.808 7761 ERROR rallytester.rallytester [-] [nova attach-volume gva_project_0...   \n",
       "28341  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "28350  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "28400  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "28478  2018-04-05 11:13:00.867 25397 ERROR rallytester.rallytester [-] [nova reboot-linux qa] Task fail...   \n",
       "28657  2018-04-05 20:01:59.527 25597 ERROR rallytester.rallytester [-] [nova attach-volume gva_project_...   \n",
       "28664  2018-04-05 21:51:49.078 26798 ERROR rallytester.rallytester [-] [nova attach-volume wig_project_...   \n",
       "28665  2018-04-05 21:51:49.078 26798 ERROR rallytester.rallytester [-] [nova attach-volume wig_project_...   \n",
       "30694  2018-04-06 03:24:42.448 23870 ERROR rallytester.rallytester [-] [nova ping-linux gva_project_035...   \n",
       "32047  2018-04-02 01:30:08.426 28739 ERROR rallytester.rallytester [-] [nova boot-from-snapshot-linux g...   \n",
       "32525  2018-04-11 09:03:28.984 11899 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig...   \n",
       "32526  2018-04-11 09:03:26.792 11507 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva...   \n",
       "32528  2018-04-11 09:03:47.607 11675 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva...   \n",
       "32555  2018-04-11 09:03:22.922 10892 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva...   \n",
       "32585  2018-04-11 09:03:27.143 12129 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux pre...   \n",
       "32586  2018-04-11 09:03:53.673 6820 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva_...   \n",
       "32587  2018-04-11 09:04:00.602 9952 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva_...   \n",
       "32588  2018-04-11 09:04:06.481 8669 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig_...   \n",
       "32626  2018-04-11 09:03:30.221 9586 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva_...   \n",
       "32627  2018-04-11 09:03:39.701 7214 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig_...   \n",
       "32628  2018-04-11 09:03:49.423 6221 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva_...   \n",
       "32630  2018-04-11 09:04:01.291 12322 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux gva...   \n",
       "32631  2018-04-11 09:04:06.266 6822 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux crit...   \n",
       "32644  2018-04-11 09:03:34.322 12603 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig...   \n",
       "32645  2018-04-11 09:03:57.825 10721 ERROR rallytester.rallytester [-] [nova boot-from-volume-linux wig...   \n",
       "32810  2018-04-10 06:10:12.352 3761 ERROR rallytester.rallytester [-] [nova ping-linux gva_project_038]...   \n",
       "33136  2018-04-04 10:40:17.039 12381 ERROR rallytester.rallytester [-] [nova reboot-linux gva_project_0...   \n",
       "\n",
       "                    dtime  \\\n",
       "9162  2018-04-02 07:00:00   \n",
       "20215 2018-04-18 10:00:00   \n",
       "27306 2018-04-05 09:00:00   \n",
       "27339 2018-04-05 09:00:00   \n",
       "27354 2018-04-05 09:00:00   \n",
       "27399 2018-04-05 09:00:00   \n",
       "27453 2018-04-05 09:00:00   \n",
       "27582 2018-04-05 09:00:00   \n",
       "27690 2018-04-05 09:00:00   \n",
       "27762 2018-04-05 09:00:00   \n",
       "27787 2018-04-05 09:00:00   \n",
       "27825 2018-04-05 09:00:00   \n",
       "27881 2018-04-05 09:00:00   \n",
       "27907 2018-04-05 09:00:00   \n",
       "27914 2018-04-05 09:00:00   \n",
       "27915 2018-04-05 09:00:00   \n",
       "27953 2018-04-05 09:00:00   \n",
       "28010 2018-04-05 09:00:00   \n",
       "28025 2018-04-05 09:00:00   \n",
       "28146 2018-04-05 17:00:00   \n",
       "28149 2018-04-05 08:00:00   \n",
       "28341 2018-04-05 09:00:00   \n",
       "28350 2018-04-05 09:00:00   \n",
       "28400 2018-04-05 09:00:00   \n",
       "28478 2018-04-05 09:00:00   \n",
       "28657 2018-04-05 18:00:00   \n",
       "28664 2018-04-05 19:00:00   \n",
       "28665 2018-04-05 19:00:00   \n",
       "30694 2018-04-06 01:00:00   \n",
       "32047 2018-04-01 23:00:00   \n",
       "32525 2018-04-11 07:00:00   \n",
       "32526 2018-04-11 07:00:00   \n",
       "32528 2018-04-11 07:00:00   \n",
       "32555 2018-04-11 07:00:00   \n",
       "32585 2018-04-11 07:00:00   \n",
       "32586 2018-04-11 07:00:00   \n",
       "32587 2018-04-11 07:00:00   \n",
       "32588 2018-04-11 07:00:00   \n",
       "32626 2018-04-11 07:00:00   \n",
       "32627 2018-04-11 07:00:00   \n",
       "32628 2018-04-11 07:00:00   \n",
       "32630 2018-04-11 07:00:00   \n",
       "32631 2018-04-11 07:00:00   \n",
       "32644 2018-04-11 07:00:00   \n",
       "32645 2018-04-11 07:00:00   \n",
       "32810 2018-04-10 04:00:00   \n",
       "33136 2018-04-04 08:00:00   \n",
       "\n",
       "                                                                                                       msg  \\\n",
       "9162    Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "20215  The server is currently unavailable. Please try again at a later time.<br /><br />    (HTTP 503)...   \n",
       "27306   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27339   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27354   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27399   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27453   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27582   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27690   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27762   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27787   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27825   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27881   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27907   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27914   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27915   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "27953   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28010   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28025   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28146   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28149   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28341   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28350   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28400   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28478   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28657   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28664   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "28665   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "30694   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "32047   Unable to establish connection to https:some_url (\\'Connection aborted.\\', BadStatusLine(\"\\'\\'\",))   \n",
       "32525                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32526                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32528                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32555                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32585                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32586                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32587                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32588                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32626                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32627                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32628                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32630                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32631                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32644                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32645                                        Build of instance xyz aborted: Service Unavailable (HTTP 503)   \n",
       "32810  Script execution failed: Error running command {\"interpreter\": \"/bin/sh\", \"script_inline\": \"ping...   \n",
       "33136                                          Connection to the hypervisor is broken on host: xyz.cern.ch   \n",
       "\n",
       "                                                                                                     _info  \\\n",
       "9162   Resource <Server: rally-e9e5-VJdA> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "20215  The server is currently unavailable. Please try again at a later time.<br /><br />    (HTTP 503)...   \n",
       "27306  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27339  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27354  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27399  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27453  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27582  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27690  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27762  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27787  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27825  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27881  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27907  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27914  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27915  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "27953  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28010  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28025  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28146  Resource <Server: rally-2150-kVPS> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28149  Resource <Server: rally-9272-P7n5> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28341  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28350  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28400  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28478  Resource <Server: rally-0d45-N5Bh> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28657  Resource <Server: rally-f1c8-hEwi> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28664  Resource <Server: rally-261f-Bn42> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "28665  Resource <Server: rally-261f-Bn42> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "30694  Resource <Server: rally-798e-7yAV> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "32047  Resource <Server: rally-cdec-Ujlg> has ERROR status. Fault: {u'message': u'Unable to establish c...   \n",
       "32525  Resource <Server: rally-66b8-K4E5> has ERROR status. Fault: {u'message': u'Build of instance 860...   \n",
       "32526  Resource <Server: rally-efaf-125N> has ERROR status. Fault: {u'message': u'Build of instance f7b...   \n",
       "32528  Resource <Server: rally-af3e-i2Yw> has ERROR status. Fault: {u'message': u'Build of instance 280...   \n",
       "32555  Resource <Server: rally-f865-WR2B> has ERROR status. Fault: {u'message': u'Build of instance a25...   \n",
       "32585  Resource <Server: rally-76ab-4IZk> has ERROR status. Fault: {u'message': u'Build of instance 5e9...   \n",
       "32586  Resource <Server: rally-1fe3-msFb> has ERROR status. Fault: {u'message': u'Build of instance 75b...   \n",
       "32587  Resource <Server: rally-8c5a-c7MX> has ERROR status. Fault: {u'message': u'Build of instance 556...   \n",
       "32588  Resource <Server: rally-1e6a-twpC> has ERROR status. Fault: {u'message': u'Build of instance eb6...   \n",
       "32626  Resource <Server: rally-e1fe-ahg2> has ERROR status. Fault: {u'message': u'Build of instance 96b...   \n",
       "32627  Resource <Server: rally-6490-afBB> has ERROR status. Fault: {u'message': u'Build of instance f91...   \n",
       "32628  Resource <Server: rally-6d6b-dOgn> has ERROR status. Fault: {u'message': u'Build of instance 5c1...   \n",
       "32630  Resource <Server: rally-0f56-eXg7> has ERROR status. Fault: {u'message': u'Build of instance 43e...   \n",
       "32631  Resource <Server: rally-186e-Tfnx> has ERROR status. Fault: {u'message': u'Build of instance 6a3...   \n",
       "32644  Resource <Server: rally-ef1c-55VR> has ERROR status. Fault: {u'message': u'Build of instance 1c0...   \n",
       "32645  Resource <Server: rally-6e4b-ocjj> has ERROR status. Fault: {u'message': u'Build of instance f87...   \n",
       "32810  Script execution failed: Error running command {\"interpreter\": \"/bin/sh\", \"script_inline\": \"ping...   \n",
       "33136  Resource <Server: rally-9f83-Kq8d> has ERROR status. Fault: {u'message': u'Connection to the hyp...   \n",
       "\n",
       "       word_count  \n",
       "9162           21  \n",
       "20215          21  \n",
       "27306          21  \n",
       "27339          21  \n",
       "27354          21  \n",
       "27399          21  \n",
       "27453          21  \n",
       "27582          21  \n",
       "27690          21  \n",
       "27762          21  \n",
       "27787          21  \n",
       "27825          21  \n",
       "27881          21  \n",
       "27907          21  \n",
       "27914          21  \n",
       "27915          21  \n",
       "27953          21  \n",
       "28010          21  \n",
       "28025          21  \n",
       "28146          21  \n",
       "28149          21  \n",
       "28341          21  \n",
       "28350          21  \n",
       "28400          21  \n",
       "28478          21  \n",
       "28657          21  \n",
       "28664          21  \n",
       "28665          21  \n",
       "30694          21  \n",
       "32047          21  \n",
       "32525          21  \n",
       "32526          21  \n",
       "32528          21  \n",
       "32555          21  \n",
       "32585          21  \n",
       "32586          21  \n",
       "32587          21  \n",
       "32588          21  \n",
       "32626          21  \n",
       "32627          21  \n",
       "32628          21  \n",
       "32630          21  \n",
       "32631          21  \n",
       "32644          21  \n",
       "32645          21  \n",
       "32810          21  \n",
       "33136          21  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The word_count with the highest frquncy is \"9\"\n",
    "filtered_data = data.query(\"word_count == 21\")\n",
    "filtered_data.head(47)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "msg_array = np.array(filtered_data._info.copy())\n",
    "# msg_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "msg_array = np.array(filtered_data._info.copy())\n",
    "jaccard_sim_set = []\n",
    "jaccard_matrix = [[0 for i in range(len(msg_array))] for j in range(len(msg_array))]\n",
    "for i in range(0, len(msg_array)):\n",
    "    jaccard_matrix[i][i] = 1\n",
    "    for j in range(i+1, len(msg_array)):\n",
    "        jaccard_sim = jaccard_similarity(msg_array[i], msg_array[j])\n",
    "        jaccard_sim_set.append(jaccard_sim)\n",
    "        jaccard_matrix[i][j] = jaccard_sim\n",
    "        jaccard_matrix[j][i] = jaccard_sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set(jaccard_sim_set)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(jaccard_sim_set))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "jaccard_dataframe = pd.DataFrame(jaccard_sim_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x120e67470>]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEICAYAAABRSj9aAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAEiBJREFUeJzt3H+s3XV9x/Hn21aU9WILVm8I7SyLdRujEeEGcS7brWym4EJJhgSH0pK6JoqLm91it/3hfiXDGCSDELUbpsWghbGxNhRnSOGGuKzMdgjlx9QrFu0dowPq1Quoq3vvj/PBXEvrOffc77nnnk+fj+Tmfr+f7+d8zufd8+3rfs/3e843MhNJUr1e0e8JSJJ6y6CXpMoZ9JJUOYNekipn0EtS5Qx6SaqcQS9JlTPopWOIiNMi4s6IeD4inoyI3+33nKRuLez3BKR56ibgR8AwcA6wKyIeysxH+zstaebCb8ZKPy0iFgGHgbMz8+ul7XPARGZu7uvkpC546kZ6uTcBR14K+eIh4Ff6NB9pVgx66eWGgO8d1TYJnNKHuUizZtBLLzcFvOaottcA3+/DXKRZM+ill/s6sDAiVk5rezPghVgNJC/GSscQEduBBN5P61M3dwO/6qduNIg8opeO7YPAycAh4AvABwx5DSqP6CWpch7RS1LlDHpJqpxBL0mVM+glqXLz4qZmS5cuzRUrVnT12Oeff55FixY1O6F5zppPDNZ8YphNzfv27XsmM1/Xrt+8CPoVK1awd+/erh47NjbG6OhosxOa56z5xGDNJ4bZ1BwRT3bSz1M3klQ5g16SKmfQS1LlDHpJqpxBL0mVM+glqXIGvSRVzqCXpMoZ9JJUuXnxzdjZ2D8xyfrNu/o9jTm1adWReV/zgWvf1e8pSCo8opekyhn0klQ5g16SKmfQS1LlDHpJqpxBL0mVM+glqXIGvSRVzqCXpMoZ9JJUuYG/BYKkzqyY57fNgMG4vUfTtq5Z1PPn6OiIPiIORMT+iPhqROwtbadFxD0R8Y3y+9TSHhFxQ0SMR8TDEXFuLwuQJP1sMzl1szozz8nMkbK+GdidmSuB3WUd4CJgZfnZCHyqqclKkmZuNufo1wLbyvI24NJp7bdkyx5gSUScPovnkSTNQmRm+04R3wIOAwl8JjO3RMR3M3NJ2R7A4cxcEhF3Addm5pfLtt3ARzNz71FjbqR1xM/w8PB527dv76qAQ89N8vSLXT10YA2fzLyvedUZixsdb2pqiqGhoUbHnO+arnn/xGRjY/XKIOzbTTtz8YKuX+fVq1fvm3aW5bg6vRj7a5k5ERGvB+6JiP+cvjEzMyLa/8X46cdsAbYAjIyM5Ojo6Ewe/hM33rqD6/afWNeUN606Mu9rPnDlaKPjjY2N0e0+MqiarnkQLnIOwr7dtK1rFvV83+7o1E1mTpTfh4A7gfOBp186JVN+HyrdJ4Dl0x6+rLRJkvqgbdBHxKKIOOWlZeCdwCPATmBd6bYO2FGWdwJXlU/fXABMZuZTjc9cktSRTt4jDQN3tk7DsxD4fGb+S0R8Bbg9IjYATwKXl/53AxcD48ALwNWNz1qS1LG2QZ+ZTwBvPkb7s8CFx2hP4JpGZidJmjVvgSBJlTPoJalyBr0kVc6gl6TKGfSSVDmDXpIqZ9BLUuUMekmqnEEvSZUz6CWpcga9JFXOoJekyhn0klQ5g16SKmfQS1LlDHpJqpxBL0mVM+glqXIGvSRVzqCXpMoZ9JJUOYNekipn0EtS5Qx6SaqcQS9JlTPoJalyBr0kVc6gl6TKGfSSVDmDXpIq13HQR8SCiHgwIu4q62dGxAMRMR4Rt0XESaX9VWV9vGxf0ZupS5I6MZMj+g8Dj09b/zhwfWa+ETgMbCjtG4DDpf360k+S1CcdBX1ELAPeBfx9WQ/gHcAdpcs24NKyvLasU7ZfWPpLkvogMrN9p4g7gL8BTgH+CFgP7ClH7UTEcuCLmXl2RDwCrMnMg2XbN4G3ZuYzR425EdgIMDw8fN727du7KuDQc5M8/WJXDx1Ywycz72tedcbiRsebmppiaGio0THnu6Zr3j8x2dhYvTII+3bTzly8oOvXefXq1fsyc6Rdv4XtOkTEbwOHMnNfRIx2NZtjyMwtwBaAkZGRHB3tbugbb93BdfvbllGVTauOzPuaD1w52uh4Y2NjdLuPDKqma16/eVdjY/XKIOzbTdu6ZlHP9+1O/kXfDlwSERcDrwZeA/wtsCQiFmbmEWAZMFH6TwDLgYMRsRBYDDzb+MwlSR1pe44+M/8kM5dl5grgCuDezLwSuA+4rHRbB+woyzvLOmX7vdnJ+SFJUk/M5nP0HwU+EhHjwGuBm0v7zcBrS/tHgM2zm6IkaTZmdDIsM8eAsbL8BHD+Mfr8AHh3A3OTJDXAb8ZKUuUMekmqnEEvSZUz6CWpcga9JFXOoJekyhn0klQ5g16SKmfQS1LlDHpJqpxBL0mVM+glqXIGvSRVzqCXpMoZ9JJUOYNekipn0EtS5Qx6SaqcQS9JlTPoJalyBr0kVc6gl6TKGfSSVDmDXpIqZ9BLUuUMekmqnEEvSZUz6CWpcga9JFWubdBHxKsj4t8j4qGIeDQi/qK0nxkRD0TEeETcFhEnlfZXlfXxsn1Fb0uQJP0snRzR/xB4R2a+GTgHWBMRFwAfB67PzDcCh4ENpf8G4HBpv770kyT1Sdugz5apsvrK8pPAO4A7Svs24NKyvLasU7ZfGBHR2IwlSTMSmdm+U8QCYB/wRuAm4BPAnnLUTkQsB76YmWdHxCPAmsw8WLZ9E3hrZj5z1JgbgY0Aw8PD523fvr2rAg49N8nTL3b10IE1fDLzvuZVZyxudLypqSmGhoYaHXO+a7rm/ROTjY3VK4OwbzftzMULun6dV69evS8zR9r1W9jJYJn5Y+CciFgC3An8Ulez+ukxtwBbAEZGRnJ0dLSrcW68dQfX7e+ojGpsWnVk3td84MrRRscbGxuj231kUDVd8/rNuxobq1cGYd9u2tY1i3q+b8/oUzeZ+V3gPuBtwJKIeOkVWQZMlOUJYDlA2b4YeLaR2UqSZqyTT928rhzJExEnA78FPE4r8C8r3dYBO8ryzrJO2X5vdnJ+SJLUE528Rzod2FbO078CuD0z74qIx4DtEfHXwIPAzaX/zcDnImIceA64ogfzliR1qG3QZ+bDwFuO0f4EcP4x2n8AvLuR2UmSZs1vxkpS5Qx6SaqcQS9JlTPoJalyBr0kVc6gl6TKGfSSVDmDXpIqZ9BLUuUMekmqnEEvSZUz6CWpcga9JFXOoJekyhn0klQ5g16SKmfQS1LlDHpJqpxBL0mVM+glqXIGvSRVzqCXpMoZ9JJUOYNekipn0EtS5Qx6SaqcQS9JlTPoJalyBr0kVa5t0EfE8oi4LyIei4hHI+LDpf20iLgnIr5Rfp9a2iMiboiI8Yh4OCLO7XURkqTj6+SI/giwKTPPAi4AromIs4DNwO7MXAnsLusAFwEry89G4FONz1qS1LG2QZ+ZT2Xmf5Tl7wOPA2cAa4Ftpds24NKyvBa4JVv2AEsi4vTGZy5J6khkZuedI1YA9wNnA9/OzCWlPYDDmbkkIu4Crs3ML5dtu4GPZubeo8baSOuIn+Hh4fO2b9/eVQGHnpvk6Re7eujAGj6ZeV/zqjMWNzre1NQUQ0NDjY453zVd8/6JycbG6pVB2LebdubiBV2/zqtXr96XmSPt+i3sdMCIGAL+EfiDzPxeK9tbMjMjovO/GK3HbAG2AIyMjOTo6OhMHv4TN966g+v2d1xGFTatOjLvaz5w5Wij442NjdHtPjKomq55/eZdjY3VK4Owbzdt65pFPd+3O/rUTUS8klbI35qZ/1San37plEz5fai0TwDLpz18WWmTJPVBJ5+6CeBm4PHM/OS0TTuBdWV5HbBjWvtV5dM3FwCTmflUg3OWJM1AJ++R3g68D9gfEV8tbX8KXAvcHhEbgCeBy8u2u4GLgXHgBeDqRmcsSZqRtkFfLqrGcTZfeIz+CVwzy3lJkhriN2MlqXIGvSRVzqCXpMoZ9JJUOYNekipn0EtS5Qx6SaqcQS9JlTPoJalyBr0kVc6gl6TKGfSSVDmDXpIqZ9BLUuUMekmqnEEvSZUz6CWpcga9JFXOoJekyhn0klQ5g16SKmfQS1LlDHpJqpxBL0mVM+glqXIGvSRVzqCXpMoZ9JJUOYNekirXNugj4rMRcSgiHpnWdlpE3BMR3yi/Ty3tERE3RMR4RDwcEef2cvKSpPY6OaLfCqw5qm0zsDszVwK7yzrARcDK8rMR+FQz05Qkdatt0Gfm/cBzRzWvBbaV5W3ApdPab8mWPcCSiDi9qclKkmau23P0w5n5VFn+b2C4LJ8BfGdav4OlTZLUJwtnO0BmZkTkTB8XERtpnd5heHiYsbGxrp5/+GTYtOpIV48dVINQc7ev5/FMTU01PuZ813TN832fgcHYt5s2F/t2t0H/dEScnplPlVMzh0r7BLB8Wr9lpe1lMnMLsAVgZGQkR0dHu5rIjbfu4Lr9s/57NVA2rToy72s+cOVoo+ONjY3R7T4yqJquef3mXY2N1SuDsG83beuaRT3ft7s9dbMTWFeW1wE7prVfVT59cwEwOe0UjySpD9r+6YyILwCjwNKIOAh8DLgWuD0iNgBPApeX7ncDFwPjwAvA1T2YsyRpBtoGfWa+5zibLjxG3wSume2kJEnN8ZuxklQ5g16SKmfQS1LlDHpJqpxBL0mVM+glqXIGvSRVzqCXpMoZ9JJUOYNekipn0EtS5Qx6SaqcQS9JlTPoJalyBr0kVc6gl6TKGfSSVDmDXpIqZ9BLUuUMekmqnEEvSZUz6CWpcgv7PQHVacXmXY2Ot2nVEdY3POZ8dyLWrN7wiF6SKmfQS1LlDHpJqpxBL0mVM+glqXIGvSRVzqCXpMr1JOgjYk1EfC0ixiNicy+eQ5LUmcaDPiIWADcBFwFnAe+JiLOafh5JUmd6cUR/PjCemU9k5o+A7cDaHjyPJKkDkZnNDhhxGbAmM99f1t8HvDUzP3RUv43AxrL6i8DXunzKpcAzXT52UFnzicGaTwyzqfkNmfm6dp36dq+bzNwCbJntOBGxNzNHGpjSwLDmE4M1nxjmouZenLqZAJZPW19W2iRJfdCLoP8KsDIizoyIk4ArgJ09eB5JUgcaP3WTmUci4kPAl4AFwGcz89Gmn2eaWZ/+GUDWfGKw5hNDz2tu/GKsJGl+8ZuxklQ5g16SKjcwQd/utgoR8aqIuK1sfyAiVsz9LJvVQc0fiYjHIuLhiNgdEW/oxzyb1OntMyLidyIiI2LgP4rXSc0RcXl5rR+NiM/P9Ryb1sG+/fMRcV9EPFj274v7Mc+mRMRnI+JQRDxynO0RETeUf4+HI+LcRieQmfP+h9ZF3W8CvwCcBDwEnHVUnw8Cny7LVwC39Xvec1DzauDnyvIHToSaS79TgPuBPcBIv+c9B6/zSuBB4NSy/vp+z3sOat4CfKAsnwUc6Pe8Z1nzrwPnAo8cZ/vFwBeBAC4AHmjy+QfliL6T2yqsBbaV5TuACyMi5nCOTWtbc2bel5kvlNU9tL6zMMg6vX3GXwEfB34wl5PrkU5q/j3gpsw8DJCZh+Z4jk3rpOYEXlOWFwP/NYfza1xm3g889zO6rAVuyZY9wJKIOL2p5x+UoD8D+M609YOl7Zh9MvMIMAm8dk5m1xud1DzdBlpHBIOsbc3lLe3yzNw1lxProU5e5zcBb4qIf42IPRGxZs5m1xud1PznwHsj4iBwN/D7czO1vpnp//cZ6dstENSciHgvMAL8Rr/n0ksR8Qrgk8D6Pk9lri2kdfpmlNa7tvsjYlVmfrevs+qt9wBbM/O6iHgb8LmIODsz/6/fExtEg3JE38ltFX7SJyIW0nq79+yczK43OrqVRET8JvBnwCWZ+cM5mluvtKv5FOBsYCwiDtA6l7lzwC/IdvI6HwR2Zub/Zua3gK/TCv5B1UnNG4DbATLz34BX07r5V616euuYQQn6Tm6rsBNYV5YvA+7NcpVjQLWtOSLeAnyGVsgP+nlbaFNzZk5m5tLMXJGZK2hdl7gkM/f2Z7qN6GTf/mdaR/NExFJap3KemMtJNqyTmr8NXAgQEb9MK+j/Z05nObd2AleVT99cAExm5lNNDT4Qp27yOLdViIi/BPZm5k7gZlpv78ZpXfS4on8znr0Oa/4EMAT8Q7nu/O3MvKRvk56lDmuuSoc1fwl4Z0Q8BvwY+OPMHNh3qx3WvAn4u4j4Q1oXZtcP8oFbRHyB1h/rpeW6w8eAVwJk5qdpXYe4GBgHXgCubvT5B/jfTpLUgUE5dSNJ6pJBL0mVM+glqXIGvSRVzqCXpMoZ9JJUOYNekir3/wYJGxN8pEyfAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jaccard_dataframe.hist(bins = 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.3125"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jaccard_matrix[-1][-3] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#      Minhashing starts here\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shingle generators \n",
    "# Arguments : Message string, shingle size {in words}\n",
    "# Returns : All shingles formed with k words\n",
    "def shingle_generator(message, k):\n",
    "    message = message.split(' ')\n",
    "    item_set = []\n",
    "    for i in range(0, len(message)-k):\n",
    "        shingle =  \"\"\n",
    "        for j in range(0, k):\n",
    "            shingle = shingle + message[i+j] + \" \"\n",
    "        item_set.append(shingle)\n",
    "    return item_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_msgs = filtered_data._info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the set of words of all raw messages\n",
    "message_set = []\n",
    "# %time\n",
    "\n",
    "for item in np.array(raw_msgs):\n",
    "#     print(item)\n",
    "    item = str(item)\n",
    "    set_of_words = set(item.split(' '))\n",
    "    # Shingle size is set to 3 words\n",
    "    set_of_words = set(shingle_generator(item, 2))\n",
    "    message_set.append(set_of_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import csv\n",
    "import pandas\n",
    "from datasketch import MinHash, MinHashLSH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "47"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate MinHash instances for all messages\n",
    "m = []\n",
    "for i in range (0, len(message_set)):\n",
    "    m_temp = MinHash(num_perm=128)\n",
    "    m.append(m_temp)\n",
    "len(m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "for message in message_set:\n",
    "    for word in message:\n",
    "        m[i].update(word.encode('utf-8'))\n",
    "    i = i + 1\n",
    "\n",
    "# Create an LSH indexb\n",
    "# %time\n",
    "lsh = MinHashLSH(threshold=.4, num_perm=128)\n",
    "count = 0\n",
    "for i in range (0, len(m)):\n",
    "    string = \"m\" + str(i)\n",
    "    lsh.insert(string, m[i])\n",
    "    count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find all the messages similar to the fist message with a specified threshold\n",
    "result = lsh.query(m[0])\n",
    "# (result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat the clustering proc for every message. \n",
    "clusters = {}\n",
    "cluster_count = 1\n",
    "# For every message add the result to the cluster\n",
    "for i in range(0, len(m)):\n",
    "    # check if the message belongs to any cluster\n",
    "    msg_str = \"m\" + str(i)\n",
    "    flag = 1\n",
    "    for item in clusters.values():\n",
    "        if msg_str in item:\n",
    "            flag = 0\n",
    "            break\n",
    "    # End of check box\n",
    "    if flag:\n",
    "        result = lsh.query(m[i])\n",
    "        clusters[cluster_count] = result\n",
    "        cluster_count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: ['m15',\n",
       "  'm17',\n",
       "  'm26',\n",
       "  'm20',\n",
       "  'm11',\n",
       "  'm28',\n",
       "  'm25',\n",
       "  'm3',\n",
       "  'm5',\n",
       "  'm16',\n",
       "  'm10',\n",
       "  'm24',\n",
       "  'm22',\n",
       "  'm14',\n",
       "  'm9',\n",
       "  'm2',\n",
       "  'm12',\n",
       "  'm21',\n",
       "  'm18',\n",
       "  'm29',\n",
       "  'm23',\n",
       "  'm8',\n",
       "  'm27',\n",
       "  'm7',\n",
       "  'm19',\n",
       "  'm6',\n",
       "  'm0',\n",
       "  'm13',\n",
       "  'm4'],\n",
       " 2: ['m1'],\n",
       " 3: ['m43',\n",
       "  'm40',\n",
       "  'm35',\n",
       "  'm38',\n",
       "  'm31',\n",
       "  'm36',\n",
       "  'm30',\n",
       "  'm39',\n",
       "  'm41',\n",
       "  'm33',\n",
       "  'm37',\n",
       "  'm34',\n",
       "  'm44',\n",
       "  'm32',\n",
       "  'm42'],\n",
       " 4: ['m45'],\n",
       " 5: ['m15',\n",
       "  'm17',\n",
       "  'm26',\n",
       "  'm20',\n",
       "  'm11',\n",
       "  'm28',\n",
       "  'm25',\n",
       "  'm3',\n",
       "  'm5',\n",
       "  'm16',\n",
       "  'm10',\n",
       "  'm24',\n",
       "  'm22',\n",
       "  'm14',\n",
       "  'm9',\n",
       "  'm2',\n",
       "  'm12',\n",
       "  'm21',\n",
       "  'm18',\n",
       "  'm29',\n",
       "  'm23',\n",
       "  'm8',\n",
       "  'm27',\n",
       "  'm7',\n",
       "  'm6',\n",
       "  'm46',\n",
       "  'm13',\n",
       "  'm4']}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
